clog/uwsgi_plugin.py

```diff
@@ -0,0 +1,58 @@
+import uwsgi
+import uwsgidecorators
+import logging
+import clog
+import clog.global_state
+import clog.handlers
+
+try:
+    import cPickle as pickle
+except ImportError:
+    import pickle
+
+
+def _mule_msg(*args, **kwargs):
+    mule = kwargs.pop('mule', 1)
+    data = pickle.dumps({
+        'service': 'uwsgi_mulefunc',
+        'func': 'log_line',
+        'args': args,
+        'kwargs': kwargs,
+    })
+    uwsgi.mule_msg(data, mule)
+
+
+class UwsgiHandler(logging.Handler):
+
+    def __init__(self, stream, mule=1):
+        logging.Handler.__init__(self)
+        self.stream = stream
+        self.mule = mule
+
+    def emit(self, record):
+        try:
+            msg = self.format(record)
+            _mule_msg(self.stream, msg, mule=self.mule)
+        except:
+            self.handleError(record)
+
+
+# TODO: Patch upstream uwsgi to return success on the pipe write so
+# we can optionally default to sync log_line
+def uwsgi_log_line(stream, line, mule=1):
+    _mule_msg(stream, line, mule=mule)
+
+
+def uwsgi_patch_global_state():
+    map(
+        lambda x: setattr(x, 'log_line', uwsgi_log_line),
+        (clog, clog.global_state)
+    )
+
+
+# Couple setup tasks at import:
+# 1. Register 'log_line' in all mules for dispatch handling on the mule side
+# 2. Insert 'UwsgiHandler' into the clog.handlers module for seamless usage
+
+uwsgidecorators.mule_functions['log_line'] = clog.global_state.log_line
+setattr(clog.handlers, 'UwsgiHandler', UwsgiHandler)
```

Comment:
I'd definitely be a fan of that, in fact my first idea for implementing this was a very simple decorator:
```python
def maybe_uwsgi(fn):
    try:
        import uwsgi
        import uwsgidecorators
    except ImportError:
        return fn

    return uwsgidecorators.mulefunc(1)(fn)


@maybe_uwsgi
def log_line...
```
But after thinking about it I wasn't sure if forcing users to use offloaded logging is a good idea (especially before I can get [this PR landed to upstream](https://github.com/unbit/uwsgi/pull/1482)), and instead thought we should try to do some kind of enabled + opt-in approach. The current PR has an associated (pending) change in `yelp_servlib.clog_util` which checks for a config that triggers the global patch such that enabling offloading everywhere is just:
```yaml
enable_uwsgi_offload_logging: true
```
in the clog module config, while alternatively letting people selectively use the `clog.uwsgi_log_line`/`clog.handlers.UwsgiHandler` if they don't want to globally patch.

Personally I don't think there's a good argument not to, outside of not running any mules and thus all logging will fail, but in that case you're already hacking the `yelp_servlib.uwsgi_serviceinit` package default arguments and theoretically know what you're doing.

/shrug I'm really not sure either, I think the current PR gives more flexibility at the cost of clarity + maintainability, but perhaps it's worth it to have the opt-in behavior. If only we didn't have to do import-time setup stuff